File: Intents/app/src/main/java/com/example/intents/MainActivity.java
package com.example.intents;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

public class MainActivity extends AppCompatActivity {

    Button btn;
    EditText edt1, edt2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        Intent intent = new Intent(MainActivity.this, SecondActivity.class);

        btn = findViewById(R.id.intent);
        edt1 = findViewById(R.id.name);
        edt2 = findViewById(R.id.qualification);

        btn.setText("To to Second Screen");


        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String title = edt1.getText().toString();
                String qualification = edt2.getText().toString();
                Log.d("Intents", title);
                intent.putExtra("title", title);
                intent.putExtra("qualification", qualification);
                intent.putExtra("Reg No.", 100221);
                startActivity(intent);
            }
        });
        Log.d("intent", "created");
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        Log.d("intent", "restarted");
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.d("intent", "paused");
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.d("intent", "started");
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.d("intent", "stopped");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d("intent", "destroyed");
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.d("intent", "resume");
    }
}

File: ViewModel/app/src/main/java/com/example/viewmodeldemo/MainActivity.kt
package com.example.viewmodeldemo

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.example.viewmodeldemo.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private lateinit var viewModel: MainActivityViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//        setContentView(R.layout.activity_main)

        Log.d("TAG", "Created")

        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        viewModel = ViewModelProvider(this)[MainActivityViewModel::class.java]

        binding.sum.text = viewModel.getSum().toString()
        binding.incrementBtn.setOnClickListener {
            val value = binding.addNumber.text.toString().toInt()
            viewModel.updateSum(value)
            binding.sum.text = viewModel.getSum().toString()
        }
    }

    override fun onStart() {
        super.onStart()
        Log.d("TAG", "Started")
    }

    override fun onPause() {
        super.onPause()
        Log.d("TAG", "Paused")
    }

    override fun onRestart() {
        super.onRestart()
        Log.d("TAG", "Restart")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d("TAG", "Destroyed")
    }

    override fun onResume() {
        super.onResume()
        Log.d("TAG", "Resume")
    }
}